digiroad2-api-oth/src/main/scala/fi/liikennevirasto/digiroad2/MunicipalityApi.scala

```diff
@@ -1,1132 +1,390 @@
-    val roadLinks = roadLinkService.getRoadsLinksFromVVH(linkIds)
-    linkIds.foreach { linkId =>
-      roadLinks.find(road => road.linkId == linkId && road.administrativeClass != State).
-        getOrElse(halt(UnprocessableEntity(s"Link id: $linkId is not valid, doesn't exist or have an administrative class 'State'.")))
-    }
-    roadLinks
-  }
-
-  private def linkIdValidation(linkIds: Set[Long], roadLinks: Seq[RoadLink]) : Unit  = {
-    linkIds.foreach { linkId =>
-      roadLinks.find(road => road.linkId == linkId && road.administrativeClass != State).
-        getOrElse(halt(UnprocessableEntity(s"Link id: $linkId is not valid, doesn't exist or have an administrative class 'State'.")))
-    }
-  }
-
-  private def validateJsonField(fieldsNotAllowed: Seq[String], body: Seq[JObject] ): Unit = {
-    val notAllowed = StringBuilder.newBuilder
-
-    fieldsNotAllowed.foreach { field =>
-      if (body.exists(bd => bd \ field != JNothing))
-        notAllowed.append(field)
-    }
-    if(notAllowed.nonEmpty)
-      halt(BadRequest("Not allow " + notAllowed + " on this type of asset"))
-  }
-
-  def extractPropertyValue(key: String, properties: Seq[AssetProperties], transformation: (String, Seq[String]) => Any):  Any = {
-    val values = properties.filter { property => property.name == key }.map(_.value)
-    transformation(key, values)
-  }
-
-  def propertyValuesToString(key: String, values: Seq[String]): String = { values.mkString }
-
-  def extractProperty(key: String, properties: Seq[AssetProperties], transformation: (String, Seq[String]) => Any):  Any = {
-    extractPropertyValue(key, properties, transformation)
-  }
-
-  def propertyValueToInt(key: String, values: Seq[String]): Seq[Int] = {
-    try {
-      values.map(_.toInt)
-    } catch {
-      case e: Exception => halt(BadRequest(s"The property values for the property with name $key are not valid."))
-    }
   }
 
-  def validateAssetProperties(assetTypeId: Int, properties: Seq[Seq[AssetProperties]]):Unit = {
-
-    if (properties.forall(_.isEmpty))
-      halt(BadRequest("Missing asset properties values"))
-
-    properties.foreach { prop =>
-      validateNameInProperty(assetTypeId, prop)
+  def validateAndInsertDataset(dataset: Dataset): Int = {
+    val assets = dataset.geoJson("features").asInstanceOf[List[Map[String, Any]]]
+    val roadlinks = dataset.roadlinks
+
+    if (assets.length != dataset.roadlinks.length) {
+      awsDao.insertDataset(dataset.datasetId, write(dataset.geoJson), write(dataset.roadlinks), DatasetStatus.FeatureRoadlinksDontMatch.value)
+    }
+    else{
+      awsDao.insertDataset(dataset.datasetId, write(dataset.geoJson), write(dataset.roadlinks), DatasetStatus.Inserted.value)
+    }
+
+    if (awsDao.getDatasetStatus(dataset.datasetId) != DatasetStatus.FeatureRoadlinksDontMatch.value) {
+      val featuresWithoutIds: Int = (roadlinks, assets).zipped.map((featureRoadlinks, feature) => {
```

Comment:
This if can be deleted and all existent inside can be moved to the else on top because at that point you already validate the match between features and road links.

Reply:
Applied this suggestion